### ISSUE TITLE
About private Vs public storage

### ISSUE THREAD

**Issue by davivel** [owncloud/android#898]
ownCloud for Android stores the local copy of the files in public storage. That allows that other apps in the device may access the downloaded files with no other special permission than access to the file system. The advantage of this approach is providing an extra way to share files with other apps. The drawback is that other apps may modify the files of the account with no barrier or control from the ownCloud app, what may result in security issues,

We are planning to start using private storage instead. The lack of access though the file system will be a limitation for regular file browsers, but there are still other means to get access to the ownCloud files from other apps. Besides, we should also implement the DocumentProvider API to provide the newest and safer. 

Our main concern about this change is how the current users will receive it. We'd like to read opinions here to get an idea.

How would you see that the app directly moves the current ownCloud folder to private storage? Would you prefer it as an optional setting, or as a mandatory option?

cc @jancborchardt @MTRichards @rperezb @tobiasKaminsky @LukeOwncloud @MMMarcy @Stoyicker and any other who wants to join the discussion.

**Comment by Xenopathic** [comment]
Most apps are restricted to seeing the Downloads folder anyway, so already in many cases one has to open up a file manager to move the downloaded oC files to Downloads so other apps can use it. For the added protection, I think it'd be a great idea to move the 'persistent storage' to a private folder, but allow for some way to copy that into Downloads in the public storage. I'm not familiar with Android internals, so DocumentProvider might do exactly that, but that's what I'd like to see :smile:

**Comment by enoch85** [comment]
I would prefer no copy at all on the local storage. Also regarding insant uploads. Maybe a .tmp at the most.

**Comment by zedascouves** [comment]
I guess I'll be the devil's advocate then. I prefer public storage, because it IS easier to deal with external applications, and wouldn't have to dig in looking for where the heck is the file. So, I'm voting for an optional setting.

**Comment by LukeOwncloud** [comment]
I often use Winamp for Android to play music directly from SD card/public storage (without adding it to the Android music library). So, I also vote for an optional setting.

Further, is there a storage limit for private data for apps? Maybe this would introduce troubles?

**Comment by ppaysant** [comment]
I'm not sure to understand all advantages / drawbacks of this problem. But access to the file could be different whether you want to read or modify / delete it. I would like instant access to my (downloaded from owncloud) files for reading, without any action. For write access, it would be nice to have optionnal setting file by file.
As the last will probably be the most complicated solution to implement, my best solution would be the actual behaviour (no control, easier use, less security).

**Comment by joeplus** [comment]
Here are some use cases what I use owncloud for at the moment:

1. Sync a password safe (and keeping it up to date) between devices and accessing it from an app through the owncloud/user@server folder on Android
2. Download images and viewing them in the gallery app
3. Download music and listen to it with different apps
4.  Sync documents and accessing them from the owncloud app

Please correct me if I am wrong but after the proposed change only number 4 would still be possible, correct? If this is the case, what other possibilities would there be for use case 1 -3. I think in general it would be a good idea to check the use cases people have and check whether they are still possible or what alternatives there would be. I guess that is what this issue is for, isn't it?

**Comment by jancborchardt** [comment]
Similar to @LukeOwncloud I often use VLC player directly on the SD card. How would that be handled with private storage and DocumentProvider?


From what other people said, it seems this would best be as a setting in the »Security« area of the settings (together with the »Passcode lock«). But it would be good to also get @tobiasKaminsky’s take on this, as well as @LukasReschke’s for security concerns.

**Comment by tjsmith3** [comment]
Could the Android App have an option for which storage source to use?  This expands the use models instead of debating one over the other.  (This approach clearly adds more work to properly enable / educate the user about the selection.)

**Comment by zedascouves** [comment]
I'm OK with having the private storage on by default, as it is more secure, and having the public storage option for the "advanced" users.

**Comment by Stoyicker** [comment]
I'm going to quote http://developer.android.com/guide/topics/data/data-storage.html

>Internal Storage
>    Store private data on the device memory.
>External Storage
>    Store public data on the shared external storage.

Therefore, if you need to keep configuration data in non-standard files like shared-prefs, keep it private, otherwise keep it public - the fact that it comes from the OC app doesn't mean that it should only be accessed by it. Therefore my vote goes for public without hesitation.
Think in the desktop - I haven't tried any of the desktop versions to be honest, but I am guessing they work in a similar fashion to Dropbox, so that I can access the files with Windows Explorer, Nautilus, or whatever file browser applies, and edit them freely from outside the OC desktop application.

This said, I would strongly like to recommend that you delay this "change". Please, freeze the Android app feature-wise and work on internal quality. There is a lot of technical debt, plus somethings like https://github.com/owncloud/android/blob/fc435f64f0449b08dd6f7af3a5c9965756593a64/src/com/owncloud/android/ui/preview/FileDownloadFragment.java#87-88 which are going to make you drown in chaos quicker than you can realize if you don't address them - Android as a platform changes fast and unpredictably, and you need to carry on with it. And I don't mean that you should always be on the bleeding edge of things, but as part of a suite that intends to be serious and robust, using a flag because a component which is years old now is showing "unexpected behavior" should be, if not fixed, at least documented before taking on other details, like the one that this issue proposes. 

(long) TL;DR - If you want to go into details like these that will take the Android app to another level, set aside manpower to work on it as a long-term project, in parallel to the rest of the suite, not once every couple of years, so that they can start by improving internal quality and then going into this. But yeah, that's none of my business :3

**Comment by tobiasKaminsky** [comment]
I also think that most of the user use the owncloud app to get a sync between mobile phone and the server. They are not using the app for itself. Common usecases would not work any longer if you change the storage to private. For example, currently downloaded images are shown in the gallery app and can be shared with this app to e.g. whatsapp...

But, as always, how do dropbox and google drive are handling this?

**Comment by bgillin58** [comment]
I have an LG G Vista which has 8gb internal memory. Keeping anything of size on the internal storage is impossible. I vote that Owncloud provide an OPTION to store in the sandbox area of the external SD card.

**Comment by MMMarcy** [comment]
I think that offering the possibility to choose is the way to go, but with public as default. In fact, as stated by others is the default in every desktop client and therefore is what users expect. Moreover private storage can be problematic in devices with small internal memory and the fact that when the application is unistalled everything is wiped out concerns me (downloading again a big load of data or losing everything after a factory reset). Therefore, I'd let users pick one on the options on the first run after instructing them on both pro and cons of each. 

Finally, I propose a third option which is public storage with encryption. This allow a somewhat compromise of the two worlds and in kinda recent devices this shouldn't even impact performance at all (well, a little) since ARM processors have AES hardware instructions. 

P.S. For completeness sake I checked the Google Drive Android application and they use private storage.

**Comment by jancborchardt** [comment]
Yes – if anything public (the current way) should be default. Using private storage can be a setting in the »Security« section at best.

**Comment by ckruetze** [comment]
Not a current user, but planing on using ownCloud on android soon.
Please make it optional where to store the data.
In an ideal world I would want per share settings for the location
Options would be:
1. public
2. private
3. no data on the device (on demand access to the cloud)

**Comment by davivel** [comment]
Thanks everybody for your thoughts.

After internal discussion, seems we will go through other way. No private storage for the short term.
